communication/src/AsyncTlsConnection.cpp

```diff
@@ -0,0 +1,390 @@
+}
+
+bool AsyncTlsConnection::verifyCertificateServer(bool preverified, boost::asio::ssl::verify_context& ctx) {
+  std::string subject(512, 0);
+  X509* cert = X509_STORE_CTX_get_current_cert(ctx.native_handle());
+  if (!cert) {
+    LOG_ERROR(logger_, "No certificate from client");
+    return false;
+  }
+  X509_NAME_oneline(X509_get_subject_name(cert), subject.data(), 512);
+  auto [valid, peer_id] = checkCertificate(cert, "client", std::string(subject), std::nullopt);
+  peer_id_ = peer_id;
+  return valid;
+}
+
+std::pair<bool, NodeNum> AsyncTlsConnection::checkCertificate(X509* receivedCert,
+                                                              std::string connectionType,
+                                                              std::string subject,
+                                                              std::optional<NodeNum> expectedPeerId) {
+  // First, perform a basic sanity test, in order to eliminate a disk read if the certificate is
+  // unknown.
+  //
+  // The certificate must have a node id, as we put it in `OU` field on creation.
+  //
+  // Since we use pinning we must know who the remote peer is.
+  // `peerIdPrefixLength` stands for the length of 'OU=' substring
+  int peerIdPrefixLength = 3;
+  std::regex r("OU=\\d*", std::regex_constants::icase);
+  std::smatch sm;
+  regex_search(subject, sm, r);
+  if (sm.length() <= peerIdPrefixLength) {
+    LOG_ERROR(logger_, "OU not found or empty: " << subject);
+    return std::make_pair(false, 0);
+  }
+
+  auto remPeer = sm.str().substr(peerIdPrefixLength, sm.str().length() - peerIdPrefixLength);
+  if (0 == remPeer.length()) {
+    LOG_ERROR(logger_, "OU empty " << subject);
+    return std::make_pair(false, 0);
+  }
+
+  NodeNum remotePeerId;
+  try {
+    remotePeerId = stoul(remPeer, nullptr);
+  } catch (const std::invalid_argument& ia) {
+    LOG_ERROR(logger_, "cannot convert OU, " << subject << ", " << ia.what());
+    return std::make_pair(false, 0);
+  } catch (const std::out_of_range& e) {
+    LOG_ERROR(logger_, "cannot convert OU, " << subject << ", " << e.what());
+    return std::make_pair(false, 0);
+  }
+
+  // If the server has been verified, check that the peers match.
+  if (expectedPeerId) {
+    if (remotePeerId != expectedPeerId) {
+      LOG_ERROR(logger_, "Peers don't match, expected: " << expectedPeerId.value() << ", received: " << remPeer);
+      return std::make_pair(false, remotePeerId);
+    }
+  }
+
+  // the actual pinning - read the correct certificate from the disk and
+  // compare it to the received one
+  namespace fs = boost::filesystem;
+  fs::path path;
+  try {
+    path = fs::path(tlsTcpImpl_.config_.certificatesRootPath) / std::to_string(remotePeerId) / connectionType /
+           std::string(connectionType + ".cert");
+  } catch (std::exception& e) {
+    LOG_FATAL(logger_, "Failed to construct filesystem path: " << e.what());
+    ConcordAssert(false);
+  }
+
+  FILE* fp = fopen(path.c_str(), "r");
```

Comment:
I know that the code does not throw exceptions but I think it might be safer to wrap the pointer in a unique_ptr.

Reply:
Agreed. Done.